logstash-core/src/main/java/org/logstash/FieldReference.java

```diff
@@ -155,13 +151,25 @@ private static FieldReference parseToCache(final CharSequence reference) {
     }
 
     private static FieldReference parse(final CharSequence reference) {
-        final String[] parts = SPLIT_PATTERN.split(reference);
-        final List<String> path = new ArrayList<>(parts.length);
-        for (final String part : parts) {
-            if (!part.isEmpty()) {
-                path.add(part.intern());
+        final ArrayList<String> path = new ArrayList<>();
+        int start = 0;
+        int end = -1;
+        for (int i = 0; i < reference.length(); ++i) {
```

Comment:
@jsvd how about simply throwing `IllegalArgumentException` on messed up data (I would also stop supporting `null` here actually and simply throw on that as well)? :)

Reply:
Interesting, usually in jruby we avoid exceptions for any kind of control flow due to their cost, so I'm a bit skeptic on that route. Taking this literally, a valid but missing reference doesn't seem to be an illegal argument and it happens too frequently with all the use of sprintf in our plugins' options. Something like `"[a]dgdfgfdgdfg[[[ff"` does seem like a good candidate for that exception though, as long as we understand how that bubbles up.

Reply:
@jsvd yea that's what I meant, I'd just throw on broken bracket sequences and `null`. I wouldn't throw just because a field doesn't exist (we don't know that in this functionality anyways), but if the `CharSequence` makes no sense I'd throw an `IllegalArgumentException` with a proper message explaining why the `CharSequence` isn't valid?

Reply:
would that bubble up from the event accessors? if so any plugin that fetches/sets on a broken reference will now get an exception instead of nil and I suspect many plugins aren't equipped to deal with this change.

Reply:
I think we can still do it, just need to properly signal that breaking change between the event API and plugins through a version bump in the logstash-core-plugin-api gem

Reply:
@jsvd agreed :) let's go for it

Reply:
ok, just set expectations, considering this a breaking change in the event api means that newer logstash-core versions will not be compatible with any plugin, and all plugins must be updated to state compatibility with the bumped logstash-core-plugin-api. Like I said, we can do it, but it has a very large impact on the ecosystem. We should likely plan this for logstash 7 together with any other breaking changes to the api we'd like to do.